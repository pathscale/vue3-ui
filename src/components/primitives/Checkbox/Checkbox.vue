<script>
/* eslint no-shadow: ["error", { "allow": ["focus"] }] -- prevent warning  'focus' is already declared in the upper scope */
import { ref, watchEffect } from 'vue'
import CheckRadioMixin from '../../../mixins/CheckRadio.js'

export default {
  name: 'VCheckbox',
  props: {
    ...CheckRadioMixin.props,
    indeterminate: Boolean,
    trueValue: {
      type: [String, Number, Boolean, Function, Object, Array],
      default: true
    },
    falseValue: {
      type: [String, Number, Boolean, Function, Object, Array],
      default: false
    }
  },
  emits: ['update:modelValue'],
  setup(props, { emit }) {
    const label = ref(null)
    const input = ref(null)
    const value = ref(props.modelValue)

    const focus = () => {
      input.value.focus()
    }

    watchEffect(() => {
      emit('update:modelValue', value.value)
    })

    watchEffect(() => {
      value.value = props.modelValue
    })

    return { label, input, value, focus }
  }
}
</script>

<template>
  <label
    class="v-checkbox checkbox"
    :class="[ size, { 'is-disabled': disabled }]"
    ref="label"
    :disabled="disabled"
    @click="focus"
    @keydown.prevent.enter="label.click()">
    <input
      v-model="value"
      :indeterminate.prop="indeterminate"
      type="checkbox"
      ref="input"
      @click.stop
      :disabled="disabled"
      :required="required"
      :name="name"
      :value="nativeValue"
      :true-value="trueValue"
      :false-value="falseValue" />
    <span class="check" :class="type" />
    <span class="control-label"><slot /></span>
  </label>
</template>
