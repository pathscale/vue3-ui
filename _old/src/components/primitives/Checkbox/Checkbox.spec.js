import { shallowMount } from '@vue/test-utils'
import Checkbox from './Checkbox.vue'

let wrapper

describe('Checkbox', () => {
  beforeEach(() => {
    wrapper = shallowMount(Checkbox)
  })

  it('is called', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('has an input checkbox', () => {
    expect(wrapper.html()).toContain('input type="checkbox"')
  })

  it('emit input event with value when value change', async () => {
    wrapper.setProps({ modelValue: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.value).toBeTruthy()

    wrapper.vm.value = false
    await wrapper.vm.$nextTick()

    const valueEmitted = wrapper.emitted()['update:modelValue'][2]
    expect(valueEmitted).toContainEqual(false)
  })

  it('method focus() gives focus to the input element', async () => {
    wrapper.vm.$refs.input.focus = jest.fn()
    wrapper.vm.focus()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.$refs.input.focus).toHaveBeenCalledTimes(1)
  })
})
