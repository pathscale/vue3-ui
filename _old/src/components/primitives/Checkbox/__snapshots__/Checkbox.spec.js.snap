// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox render correctly 1`] = `<label class="v-checkbox checkbox"><input type="checkbox" value="" true-value="true" false-value="false"><span class="check"></span><span class="control-label"></span></label>`;
